File: src/api/theme.js
import { api, getAccessToken } from "./request";

const STORAGE_KEY = "aozora-theme";

export const getThemeConfig = async () => {
  try {
    if (getAccessToken()) {
      const serverConfig = await api.get("/theme");
      if (serverConfig) {
        const config = {
          palette: serverConfig.palette || "aozora",
          backgroundEnabled: serverConfig.background_enabled || false,
          backgroundImage: serverConfig.background_image || "",
          blur: serverConfig.blur || 0,
          brightness: serverConfig.brightness || 100,
          maskOpacity: serverConfig.mask_opacity || 34,
          backgroundSize: serverConfig.background_size || "cover",
          backgroundPosition: serverConfig.background_position || "center",
          cardOpacity: serverConfig.card_opacity || 72,
          dockImage: serverConfig.dock_image || "",
          glassBackground: serverConfig.glass_background || false,
        };
        localStorage.setItem(STORAGE_KEY, JSON.stringify(config));
        return config;
      }
    }
  } catch (error) {
    console.warn("从服务器获取主题配置失败，使用本地缓存", error);
  }

  const raw = localStorage.getItem(STORAGE_KEY);
  return raw ? JSON.parse(raw) : {};
};

export const saveThemeConfig = async (config) => {
  try {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(config));
  } catch (error) {
    throw new Error("背景图片过大，已超出浏览器本地存储限制", { cause: error });
  }

  const serverConfig = {
    palette: config.palette,
    background_enabled: config.backgroundEnabled,
    background_image: config.backgroundImage,
    blur: config.blur,
    brightness: config.brightness,
    mask_opacity: config.maskOpacity,
    background_size: config.backgroundSize,
    background_position: config.backgroundPosition,
    card_opacity: config.cardOpacity,
    dock_image: config.dockImage,
    glass_background: config.glassBackground,
  };

  return api.post("/theme", serverConfig);
};

File: backend/internal/models/theme.go
package models

import "gorm.io/gorm"

type ThemeConfig struct {
	BaseModel
	Palette            string `json:"palette" gorm:"default:aozora"`
	BackgroundEnabled  bool   `json:"background_enabled" gorm:"default:false"`
	BackgroundImage    string `json:"background_image" gorm:"type:text"`
	Blur               int    `json:"blur" gorm:"default:0"`
	Brightness         int    `json:"brightness" gorm:"default:100"`
	MaskOpacity        int    `json:"mask_opacity" gorm:"default:34"`
	BackgroundSize     string `json:"background_size" gorm:"default:cover"`
	BackgroundPosition string `json:"background_position" gorm:"default:center"`
	CardOpacity        int    `json:"card_opacity" gorm:"default:72"`
	// DockImage 是侧边栏左下角时钟卡的背景长图（base64 或 URL），随主题配置保存。
	DockImage string `json:"dock_image" gorm:"type:text"`
}

func (t *ThemeConfig) AfterFind(tx *gorm.DB) (err error) {
	if t.Palette == "" {
		t.Palette = "aozora"
	}
	if t.BackgroundSize == "" {
		t.BackgroundSize = "cover"
	}
	if t.BackgroundPosition == "" {
		t.BackgroundPosition = "center"
	}
	return
}
